File: python-api/acquire_trigger_software_keep_arm_in_loop.py
#!/usr/bin/python3

"""
Red Pitaya C API example - Acquiring a signal with software trigger in loop
Using Keep Arm mode to maintain acquisition between triggers
Ported to Python
"""

import time
import sys
import numpy as np
from rp_overlay import overlay
import rp

def main():
    # Initialize Red Pitaya
    fpga = overlay()

    if rp.rp_Init() != rp.RP_OK:
        print("Rp api init failed!", file=sys.stderr)
        return -1

    # Uncomment if you need debug registers
    # rp.rp_EnableDebugReg()

    # Configure generator
    print("Configuring generator...")
    rp.rp_GenReset()
    rp.rp_GenFreq(rp.RP_CH_1, 100.0)       # 100 Hz frequency
    rp.rp_GenAmp(rp.RP_CH_1, 1.0)           # 1 V amplitude
    rp.rp_GenWaveform(rp.RP_CH_1, rp.RP_WAVEFORM_SINE)
    rp.rp_GenOutEnable(rp.RP_CH_1)
    rp.rp_GenTriggerOnly(rp.RP_CH_1)

    # Buffer size
    N = 20
    arr_f = np.zeros(N, dtype=np.float32)

    # Configure acquisition
    rp.rp_AcqReset()
    rp.rp_AcqSetDecimation(rp.RP_DEC_16384)  # High decimation for slow sampling
    rp.rp_AcqSetTriggerDelay(0)               # No trigger delay

    # Enable Keep Arm mode - acquisition stays armed after trigger
    print("Enabling Keep Arm mode...")
    rp.rp_AcqSetArmKeep(True)

    # Start acquisition once (keeps running due to Keep Arm)
    rp.rp_AcqStart()
    print("Acquisition started in Keep Arm mode\n")

    # Number of acquisition cycles
    count = 10

    for step in range(1, count + 1):
        print(f"Step {step}")

        # Wait for pre-trigger buffer to fill
        print("Waiting for pre-trigger buffer...")
        for _ in range(5):
            ret, pre_counter = rp.rp_AcqGetPreTriggerCounter()
            if ret == rp.RP_OK:
                print(f"  Pre-trigger counter: {pre_counter}")
            time.sleep(1)

        # Additional delay
        time.sleep(1)

        # Set software trigger
        print("Setting software trigger...")
        rp.rp_AcqSetTriggerSrc(rp.RP_TRIG_SRC_NOW)

        # Wait for trigger to occur
        print("Waiting for trigger...")
        while True:
            ret, state = rp.rp_AcqGetTriggerState()
            if ret == rp.RP_OK and state == rp.RP_TRIG_STATE_TRIGGERED:
                print("Trigger detected!")
                time.sleep(1)  # Extra delay after trigger
                break
            time.sleep(0.01)  # Small delay to reduce CPU usage

        # Wait for buffer to fill completely
        print("Waiting for buffer to fill...")
        while True:
            ret, fill_state = rp.rp_AcqGetBufferFillState()
            if ret == rp.RP_OK and fill_state:
                print("Buffer is full!")
                break
            time.sleep(0.01)

        # Get oldest data using NumPy array
        print(f"Acquiring data (buffer size: {N})...")
        res = rp.rp_AcqGetOldestDataVNP(rp.RP_CH_1, arr_f)

        if res == rp.RP_OK:
            print("\nAcquired data:")
            for i, value in enumerate(arr_f):
                print(f"{value:.6f}")
        else:
            print(f"Error acquiring data: {rp.rp_GetError(res)}",
                  file=sys.stderr)
            break

        # Unlock trigger for Keep Arm mode to re-arm for next acquisition
        # In normal mode, unlocking occurs when calling rp_AcqStart
        print("Unlocking trigger for next acquisition...")
        rp.rp_AcqUnlockTrigger()

        print("-" * 50)

    # Stop acquisition
    rp.rp_AcqStop()

    # Release resources
    rp.rp_Release()
    print("\nProgram finished successfully")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: C/acquire_trigger_software_keep_arm_in_loop.cpp
/* Red Pitaya C API example Acquiring a signal from a buffer
 * This application acquires a signal on a specific channel */

#include "rp.h"
#include "rp_hw_calib.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char **argv) {

  /* Print error, if rp_Init() function failed */
  if (rp_Init() != RP_OK) {
    fprintf(stderr, "Rp api init failed!\n");
  }
  //  rp_EnableDebugReg();
  rp_GenReset();
  rp_GenFreq(RP_CH_1, 100.0);
  rp_GenAmp(RP_CH_1, 1.0);
  rp_GenWaveform(RP_CH_1, RP_WAVEFORM_SINE);
  rp_GenOutEnable(RP_CH_1);
  rp_GenTriggerOnly(RP_CH_1);

  uint32_t buff_size = 20;
  float *buff = (float *)malloc(buff_size * sizeof(float));

  rp_AcqReset();
  rp_AcqSetDecimation(RP_DEC_16384);
  rp_AcqSetTriggerDelay(0);

  int count = 10;
  rp_AcqSetArmKeep(true);
  rp_AcqStart();
  while (count--) {
    printf("Step %d\n", 10 - count);

    /* After acquisition is started some time delay is needed in order to
     * acquire fresh samples in to buffer*/
    /* Here we have used time delay of one second but you can calculate exact
     * value taking in to account buffer*/
    /*length and smaling rate*/
    int z = 5;

    while (z) {
      uint32_t c;
      rp_AcqGetPreTriggerCounter(&c);
      sleep(1);
      z--;
      printf("Pre counter %d\n", c);
    }

    sleep(1);

    rp_AcqSetTriggerSrc(RP_TRIG_SRC_NOW);
    rp_acq_trig_state_t state = RP_TRIG_STATE_TRIGGERED;

    while (1) {
      rp_AcqGetTriggerState(&state);
      if (state == RP_TRIG_STATE_TRIGGERED) {
        sleep(1);
        break;
      }
    }

    bool fillState = false;
    while (!fillState) {
      rp_AcqGetBufferFillState(&fillState);
    }

    rp_AcqGetOldestDataV(RP_CH_1, &buff_size, buff);

    uint32_t i;
    for (i = 0; i < buff_size; i++) {
      printf("%f\n", buff[i]);
    }

    // Unlocking the trigger is required for Keep Arm mode.
    // For normal mode, unlocking occurs when calling the rp_AcqStart command.
    rp_AcqUnlockTrigger();
  }

  rp_AcqStop();
  /* Releasing resources */
  free(buff);
  rp_Release();

  return 0;
}
